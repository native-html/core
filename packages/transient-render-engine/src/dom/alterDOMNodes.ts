import {
  Node as DOMNode,
  Text as DOMText,
  Element as DOMElement
} from 'domhandler';
import { ElementType } from 'domelementtype';

import map from 'ramda/src/map';
import identity from 'ramda/src/identity';
import always from 'ramda/src/always';
import pipe from 'ramda/src/pipe';
import reject from 'ramda/src/reject';
export interface AlterDOMParams {
  ignoreDOMNode?: (node: DOMNode) => boolean;
  alterDOMData?: (node: DOMText) => string | false | void;
  alterDOMChildren?: (node: DOMElement) => DOMNode[] | false | void;
  alterDOMElement?: (node: DOMElement) => DOMElement | false | void;
}

export default function alterDOMNodes(
  nodes: DOMNode[],
  params: AlterDOMParams
): DOMNode[] {
  const {
    ignoreDOMNode = always(false),
    alterDOMElement = identity,
    alterDOMChildren = identity,
    alterDOMData = (node) => node.data
  } = params;
  const alterNodes = pipe<DOMNode[], DOMNode[], DOMNode[]>(
    reject(ignoreDOMNode),
    map(function applyParams(node: DOMNode) {
      let nextNode = node;
      if (node.type === ElementType.Tag) {
        const elementNode =
          alterDOMElement(node as DOMElement) || (nextNode as DOMElement);
        const alteredChildren = alterDOMChildren(elementNode);
        elementNode.children = Array.isArray(alteredChildren)
          ? alteredChildren
          : elementNode.children;
        elementNode.children = alterNodes(elementNode.children);
        nextNode = elementNode;
      } else if (node.type === ElementType.Text) {
        const textNode = node as DOMText;
        const alteredData = alterDOMData(textNode);
        textNode.data =
          typeof alteredData === 'string' ? alteredData : textNode.data;
      }
      return nextNode;
    })
  );
  return alterNodes(nodes);
}
