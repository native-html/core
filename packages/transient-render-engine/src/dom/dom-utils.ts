import { Text, Element, Node, Document, NodeWithChildren } from 'domhandler';
import { Text as TextType, Tag as TagType } from 'domelementtype';

export { Text, Element, Node, NodeWithChildren, Document };

export function isDomText(node: any): node is Text {
  return node && node.type === TextType;
}

export function isDomElement(node: any): node is Element {
  return node && node.type === TagType;
}
