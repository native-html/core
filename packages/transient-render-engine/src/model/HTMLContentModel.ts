/**
 * The **content model** associated with a tag determines how this tag should
 * be translated in the Transient Render Tree.
 */
// eslint-disable-next-line no-shadow
export enum HTMLContentModel {
  /**
   * Translatable to TBlock.
   */
  block = 'block',
  /**
   * Translatable to TPhrasing and TText
   */
  textual = 'textual',
  /**
   * Translatable to TBlock, TPhrasing and TText
   */
  mixed = 'mixed',
  /**
   * Translatable to TEmpty
   */
  none = 'none'
}

export default HTMLContentModel;
