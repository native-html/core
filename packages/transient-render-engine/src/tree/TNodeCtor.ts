import { TStyles } from '../styles/TStyles';
import markersProtype from './markersProtype';
import tnodeToString from './tnodeSnapshot';
import {
  Markers,
  TNodeDescriptor,
  TNodeImpl,
  TNodeInit,
  TNodeType
} from './tree-types';

export type GenericTNodeCtor<Init = TNodeInit, Impl = TNodeImpl> = {
  new (init: Init): Impl;
  prototype: Impl;
};

export type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

function updateNodeIndexes(node: Mutable<TNodeImpl>, i: number) {
  node.nodeIndex = i;
}

const emptyAttrs = Object.freeze({});
const emptyClasses = Object.freeze([]);

function setMarkersFromAttributes(
  targetMarkers: Markers,
  _parentMarkers: Readonly<Markers>,
  { attributes }: TNodeDescriptor
): void {
  if ('dir' in attributes) {
    targetMarkers.direction = attributes.dir as 'rtl';
  }
  if ('lang' in attributes) {
    targetMarkers.lang = attributes.lang;
  }
}

const prototype: Omit<TNodeImpl, 'displayName' | 'type'> = {
  children: Object.freeze([]) as any,
  init: Object.freeze({}) as any,
  classes: Object.freeze([]) as any,
  styles: Object.freeze(TStyles.empty()) as any,
  __nodeIndex: null,
  __trimmedLeft: false,
  __trimmedRight: false,
  markers: markersProtype,
  get attributes() {
    return this.domNode?.attribs || emptyAttrs;
  },

  get hasWhiteSpaceCollapsingEnabled() {
    return typeof this.styles.webTextFlow.whiteSpace === 'string'
      ? this.styles.webTextFlow.whiteSpace === 'normal'
      : true;
  },

  get contentModel() {
    return this.elementModel?.contentModel || null;
  },

  get parentStyles() {
    return this.init.parentStyles || this.parent?.styles || null;
  },

  get id() {
    return this.attributes.id || null;
  },

  get domNode() {
    return this.init.domNode || null;
  },

  get elementModel() {
    return this.init.elementModel;
  },

  get tagName() {
    return this.init.domNode?.name || null;
  },

  get parent() {
    return (this.init.parent as any) || null;
  },

  get nodeIndex() {
    if (this.__nodeIndex === null) {
      this.__nodeIndex = this.init.nodeIndex || 0;
    }
    return this.__nodeIndex;
  },

  get isUnregistered() {
    return this.init.isUnregistered || false;
  },

  set nodeIndex(nodeIndex: number) {
    this.__nodeIndex = nodeIndex;
  },

  bindChildren(children, shouldUpdateNodeIndexes = false) {
    //@ts-ignore
    this.children = children;
    if (shouldUpdateNodeIndexes) {
      children.forEach(updateNodeIndexes);
    }
  },

  cloneInitParams(partial) {
    return Object.assign({}, this.init, partial) as any;
  },

  isCollapsibleLeft() {
    if (this.children.length) {
      return (
        this.hasWhiteSpaceCollapsingEnabled &&
        this.children[0].isCollapsibleLeft()
      );
    }
    return false;
  },

  isCollapsibleRight() {
    if (this.children.length) {
      return (
        this.hasWhiteSpaceCollapsingEnabled &&
        this.children[this.children.length - 1].isCollapsibleRight()
      );
    }
    return false;
  },

  isEmpty() {
    return false;
  },

  trimLeft() {
    if (!this.__trimmedLeft && this.children.length) {
      const firstChild = this.children[0];
      firstChild.trimLeft();
      if (firstChild.isEmpty()) {
        //@ts-ignore
        this.children.splice(0, 1);
      }
      this.__trimmedLeft = true;
    }
  },

  trimRight() {
    if (!this.__trimmedRight && this.children.length) {
      const lastChild = this.children[this.children.length - 1];
      lastChild.trimRight();
      if (lastChild.isEmpty()) {
        //@ts-ignore
        this.children.splice(-1, 1);
      }
      this.__trimmedRight = true;
    }
  },

  matchContentModel() {
    return false;
  },

  spliceChildren(indexesToSplice) {
    let offset = 0;
    for (const i of indexesToSplice) {
      //@ts-ignore
      this.children.splice(i - offset, 1);
      offset += 1;
    }
  },

  collapse() {
    this.collapseChildren();
    this.bindChildren(this.children, true);
  },

  getNativeStyles() {
    return {
      ...this.styles.nativeBlockFlow,
      ...this.styles.nativeBlockRet,
      ...this.styles.nativeTextFlow,
      ...this.styles.nativeTextRet
    };
  },

  getWebStyles() {
    return {
      ...this.styles.webTextFlow,
      ...this.styles.webBlockRet
    } as any;
  },

  collapseChildren() {
    return;
  },

  snapshot(options = {}) {
    const { withStyles = false, withNodeIndex = false } = options;
    return tnodeToString(this as any, { withStyles, withNodeIndex });
  },

  hasClass(className) {
    return this.classes.indexOf(className) > -1;
  },

  toString() {
    return this.snapshot();
  },

  setMarkers(targetMarkers, parentMarkers) {
    if (this.elementModel?.setMarkersForTNode) {
      this.elementModel.setMarkersForTNode(targetMarkers, parentMarkers, this);
    }
    setMarkersFromAttributes(targetMarkers, parentMarkers, this);
    this.init.context.setMarkersForTNode?.(targetMarkers, parentMarkers, this);
  },

  initialize<Impl extends TNodeImpl<TNodeInit> = TNodeImpl>(
    this: Mutable<Impl>,
    init: Impl['init']
  ) {
    this.init = init;
    this.classes = this.attributes.class?.split(/\s+/) || emptyClasses;
    const parentMarkers = init.parent
      ? init.parent.markers
      : markersProtype.extend();
    this.markers = parentMarkers.extend();
    this.setMarkers(this.markers, parentMarkers, this);
    // Avoid very long prototype chains by assigning parent to current
    // when current has no own properties
    if (Object.keys(this.markers).length === 0) {
      this.markers = parentMarkers;
    }
    this.styles =
      init.styles ||
      init.context.stylesMerger.buildStyles(
        this.attributes.style,
        this.parentStyles || null,
        this
      );
  }
};

const TNodeCtor = function TNode<Impl extends TNodeImpl = TNodeImpl>(
  this: Mutable<TNodeImpl>,
  type: TNodeType,
  displayName: string,
  extraAccessors?: {
    [k in Exclude<keyof Impl, keyof TNodeImpl>]: {
      get: () => any;
      set?: (val: any) => void;
    };
  }
) {
  this.type = type;
  this.displayName = displayName;
  extraAccessors && Object.defineProperties(this, extraAccessors);
};

TNodeCtor.prototype = prototype as any;

export default TNodeCtor;
