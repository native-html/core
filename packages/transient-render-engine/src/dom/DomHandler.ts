import {
  DomHandler as OriginalDomHandler,
  DomHandlerOptions as OriginalDomHandlerOptions,
  Document,
  Element,
  Node,
  Text,
  NodeWithChildren
} from 'domhandler';
import { isDomElement, isDomText } from './dom-utils';

/**
 * A record of callback to visit the DOM.
 *
 * @public
 */
export interface DomVisitorCallbacks {
  /**
   * A callback invoked when encountering a {@link Document}.
   *
   * @param document - The document to visit.
   *
   */
  onDocument?(document: Document): void;
  /**
   * A callback invoked when encountering an {@link Element}.
   *
   * @param element - The element to visit.
   */
  onElement?(element: Element): void;
  /**
   * A callback invoked when encountering a {@link Text} node.
   *
   * @param text - The text to visit.
   */
  onText?(text: Text): void;
}

export interface DomHandlerOptions extends OriginalDomHandlerOptions {
  ignoredTags?: string[];
  visitors?: DomVisitorCallbacks;
  ignoreNode?: (
    node: Node,
    parent: NodeWithChildren
  ) => boolean | void | unknown;
}

export default class DomHandler extends OriginalDomHandler {
  private ignoredTags: Record<string, boolean>;
  private ignoredTagsCount = -1;
  private visitors: DomVisitorCallbacks;

  constructor(options: DomHandlerOptions) {
    super(undefined, options, undefined);
    let ignoredTags: DomHandler['ignoredTags'] = {};
    for (const tag of options.ignoredTags || []) {
      ignoredTags[tag] = true;
    }
    this.ignoredTags = ignoredTags;
    this.visitors = options.visitors || {};
    const isIgnored = this.isIgnored.bind(this);
    if (options.ignoreNode) {
      this.isIgnored = (node) => {
        return (
          isIgnored(node) ||
          options.ignoreNode!(node, this.tagStack[this.tagStack.length - 1])
        );
      };
    }
  }

  isIgnored(node: Node): boolean | unknown | void {
    return (
      (this.ignoredTags[(node as Element).name] as boolean) ||
      this.ignoredTagsCount > -1
    );
  }

  addNode(node: Node): void {
    if (this.isIgnored(node) === true) {
      // increment only for elements.
      if (isDomElement(node)) {
        this.ignoredTagsCount++;
      }
    } else {
      super.addNode(node);
      if (isDomText(node)) {
        this.visitors.onText?.(node);
      }
    }
  }

  ontext(text: string) {
    // Only include text when out of an
    // ignoring session.
    if (this.ignoredTagsCount < 0) {
      super.ontext(text);
    }
  }

  onopentag(name: string, attribs: any): void {
    super.onopentag(name, attribs);
  }

  onclosetag() {
    // Call onElement on close to give access to this node children
    this.visitors.onElement?.(
      this.tagStack[this.tagStack.length - 1] as Element
    );
    super.onclosetag();
    if (this.ignoredTagsCount > -1) {
      this.ignoredTagsCount--;
    }
  }

  onend(): void {
    this.visitors.onDocument?.(this.root);
  }
}
