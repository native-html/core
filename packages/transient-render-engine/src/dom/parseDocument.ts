import { Document } from 'domhandler';
import { Parser, ParserOptions } from 'htmlparser2';
import DomHandler, { DomHandlerOptions } from './DomHandler';
/**
 * Parses the data, returns the resulting document.
 *
 * @param data The data that should be parsed.
 * @param options Optional options for the parser and DOM builder.
 */
export default function parseDocument(
  data: string,
  options: ParserOptions & DomHandlerOptions = {}
): Document {
  const handler = new DomHandler(options);
  new Parser(handler, options).end(data);
  return handler.root;
}
