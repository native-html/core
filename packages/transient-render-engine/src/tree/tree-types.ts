import { Element, Text } from '../dom/dom-utils';
import HTMLContentModel from '../model/HTMLContentModel';
import HTMLElementModel from '../model/HTMLElementModel';
import { TStylesShape } from '../styles/TStyles';
import { TStylesMerger } from '../styles/TStylesMerger';

export interface TNodeInit {
  readonly elementModel: HTMLElementModel<string, HTMLContentModel> | null;
  readonly context: {
    readonly stylesMerger: TStylesMerger;
    readonly setMarkersForTNode?: SetMarkersForTNode;
    readonly removeLineBreaksAroundEastAsianDiscardSet: boolean;
  };
  readonly parent?: TNodeImpl | null;
  readonly domNode?: Element | null;
  readonly styles?: TStylesShape | null;
  readonly parentStyles?: TStylesShape | null;
  readonly nodeIndex?: number;
  readonly isUnregistered?: boolean;
}

/**
 * A minimal TNode shape accessible at instantiation.
 *
 * @public
 */
export interface TNodeDescriptor {
  tagName: string | null;
  classes: string[];
  id: string | null;
  attributes: Record<string, string>;
  markers: Markers;
}

/**
 * A function to set markers for a specific {@link TNode}.
 *
 * @param targetMarkers - The markers instance to alter.
 * @param parentMarkers - Th direct ascendant markers instance for this tnode.
 * @param tnode - The minimalistic {@link TNodeDescriptor} to which the
 * `targetMarkers` will be attached.
 *
 * @public
 */
export type SetMarkersForTNode = (
  targetMarkers: Markers,
  parentMarkers: Readonly<Markers>,
  tnode: TNodeDescriptor
) => void;

/**
 * Markers form an abstraction in which one node provides semantic information
 * to itself and all its descendants. For example, `ins` elements, which stand
 * for "insertion" of content in the context of an edit will provide the {
 * edits: 'ins' } marker to all its descendants.
 *
 * @remarks
 * Custom renderers can use markers to change their layout and convey their
 * semantic meaning. Markers can be derived from attributes, such as `lang` and
 * `dir` attributes, or tag names, such as `a`, `ins`, `del`...
 *
 * **TypeScript users**: You can add fields to the {@link Markers} interface via
 * {@link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation | module augmentation}:
 *
 * ```ts
 * declare module '@native-html/transient-render-engine' {
 *   interface Markers {
 *     customField?: boolean;
 *   }
 * }
 * ```
 */
export interface Markers {
  /**
   * If this node is an `a` or has one as ancestor, this field will be set to
   * `true`.
   *
   * @defaultValue false
   */
  anchor: boolean;
  /**
   * If this node is a `del` or `ins` or has either as ancestor, this field
   * will be set accordingly. Otherwise, it will be set to 'none'.
   *
   * https://html.spec.whatwg.org/#edits
   *
   * @defaultValue 'none'
   */
  edits: 'ins' | 'del' | 'none';
  /**
   * The direction for this content. Follows `dir` attribute.
   *
   * https://html.spec.whatwg.org/#the-dir-attribute
   *
   * @defaultValue 'ltr'
   */
  direction: 'ltr' | 'rtl';
  /**
   * The language for this content. Follows `lang` attribute.
   *
   * https://html.spec.whatwg.org/#the-lang-and-xml:lang-attributes
   */
  lang: string;
  /**
   * * -1: this node is not an `ol` and has no `ol` parents
   * * 0: this node is an `ol` or has one `ol` parent
   * * 1: ...
   */
  olNestLevel: number;
  /**
   * * -1: this node is not an `ul` and has no `ul` parents
   * * 0: this node is an `ul` or has one `ul` parent
   * * 1: ...
   */
  ulNestLevel: number;
  /**
   * Create a new marker instance which extends this markers.
   */
  extend(this: Markers): Markers;
  /**
   * Return a string represenation of this marker, including inherited
   * properties from parent markers.
   */
  toString(): string;
}

/**
 * @public
 */
export interface TNodePrintOptions {
  /**
   * Include styles in snapshot.
   */
  withStyles: boolean;
  /**
   * Include node index in snapshot.
   */
  withNodeIndex: boolean;
}

/**
 * Processed styles which can be passed to `View` elements.
 *
 * @public
 */
export type NativeBlockStyles = TStylesShape['nativeBlockFlow'] &
  TStylesShape['nativeBlockRet'];

/**
 * Processed styles which can be passed to `Text` elements.
 *
 * @public
 */
export type NativeTextStyles = TStylesShape['nativeBlockFlow'] &
  TStylesShape['nativeBlockRet'] &
  TStylesShape['nativeTextFlow'] &
  TStylesShape['nativeTextRet'];

/**
 * Processed Web text styles.
 *
 * @public
 */
export type WebTextStyles = TStylesShape['webTextFlow'];

/**
 * Processed Web block styles.
 *
 * @public
 */
export type WebBlockStyles = TStylesShape['webBlockRet'];

/**
 * Shared properties for all {@link TNode} types.
 *
 * @public
 */
export interface TNodeShape<T extends TNodeType> {
  /**
   * Attributes for this tag.
   */
  readonly attributes: Record<string, string>;
  /**
   * The id for this node, extracted from the id attribute of the
   * underlying {@link Node}.
   */
  readonly id: string | null;
  /**
   * A list of classes for this node, extracted from the class attribute of the
   * underlying {@link Node}.
   */
  readonly classes: string[];
  /**
   * Non-anonymous nodes will hold a reference to a DOM node.
   */
  readonly domNode: Element | null;
  /**
   * The tag name for this node.
   *
   * @remarks Anonymous nodes generated during hoisting won't have a tag name.
   * Also, some TText nodes don't have a tagName.
   */
  readonly tagName: string | null;
  /**
   * The parent of this node before hoisting.
   *
   * @remarks
   * "Before hoisting" implies that this parent will not match "anonymous"
   * parents.
   */
  readonly parent: TDocument | TBlock | TPhrasing | null;
  /**
   * The position of this element relatively to its parents, before hoisting,
   * after collapsing.
   *
   * @remarks
   * "Before hoisting" implies that this index corresponds to the node position
   * in the DOM, after removal of empty elements as per whitespace collapsing
   * algorithm.
   */
  readonly nodeIndex: number;
  /**
   * Children of this node.
   */
  readonly children: ReadonlyArray<TNode>;

  /**
   * Used for debugging purposes.
   */
  readonly displayName: string;

  /**
   * The type of this tnode.
   */
  readonly type: T;

  /**
   * `true` when this tag is not a valid HTML tag **and** there is no custom
   * renderer for this tag.
   */
  readonly isUnregistered: boolean;

  /**
   * Markers form an abstraction in which one node provides semantic information
   * to itself and all its descendants. For example, `ins` elements, which stand
   * for "insertion" of content in the context of an edit will provide the {
   * edits: 'ins' } marker to all its descendants.
   *
   * @remarks This attribute must be considered immutable. Never try to
   * change it by hand, or you might update the markers of an anscestor! For
   * performance reasons, markers instances might be shared from parent to
   * children when they are equal.
   */
  readonly markers: Readonly<Markers>;

  /**
   * TStyles for this TNode. You should not use these unless required.
   * Use {@link TNodeShape.getNativeStyles} instead.
   */
  readonly styles: TStylesShape;

  /**
   * Create a JSX string representation of this node and its children.
   *
   * @remarks The snapshot is _just_ a representation. For example, it will
   * print `href` as an attribute of the TNode, while if you want to access
   * `href` programatically, you'll need to access it via
   * `tnode.attributes.href`.
   *
   * @param options - Customize what should be displayed.
   */
  snapshot(options?: Partial<TNodePrintOptions>): string;

  /**
   * Check if this {@link TNode} has the given `className`.
   *
   * @param className - The class to check.
   */
  hasClass(className: string): boolean;

  /**
   * Test if the given content model matches this TNode content model.
   *
   * @param contentModel - The content model to test against.
   */
  matchContentModel(contentModel: HTMLContentModel): boolean;

  /**
   * Get own native styles.
   */
  getNativeStyles(): T extends 'text' | 'phrasing'
    ? NativeTextStyles
    : NativeBlockStyles;

  /**
   * Get styles that cannot be handled by React Native.
   */
  getWebStyles(): WebTextStyles & WebBlockStyles;
}

/**
 * Metainformation extracted from `<head>` tag.
 *
 * @public
 */
export interface DocumentContext {
  charset: string;
  baseHref: string;
  baseTarget: '_blank' | '_self' | '_parent' | '_top';
  lang: string;
  dir: 'ltr' | 'rtl';
  title: string;
  meta: { name: string; value: string }[];
  links: Record<string, string>[];
}

/**
 * A transient render node (TNode) is a ready-to-render data structure.
 */
export type TNode = TBlock | TDocument | TEmpty | TPhrasing | TText;

/**
 * Transient render nodes to be rendered in React Native `Views`.
 */
export interface TBlock extends TNodeShape<'block'> {
  readonly tagName: string;
  readonly domNode: Element;
}

/**
 * The root of a transient render tree. It has a special
 * `context` field holding metadata about the page.
 */
export interface TDocument extends TNodeShape<'document'> {
  /**
   * An object containing special information for this document, such as lang,
   * dir, charset, baseHref...
   */
  readonly context: Readonly<DocumentContext>;
}

/**
 * Transient render nodes for a non-renderable elements, e.g.
 * for elements which have an {@link HTMLContentModel} set to **none**.
 */
export interface TEmpty extends TNodeShape<'empty'> {
  readonly domNode: Element;
}

/**
 * Transient render nodes to be rendered in React Native `Text`
 * components. TPhrasing nodes have {@link TText}, {@link TPhrasing} or
 * {@link TEmpty} children.
 */
export interface TPhrasing extends TNodeShape<'phrasing'> {}

/**
 * Transient render nodes to be rendered in React Native `Text`
 * components. They don't have children, and have a special `data` field
 * holding the text to be rendered.
 */
export interface TText extends TNodeShape<'text'> {
  readonly data: string;
  readonly textNode: Text;
}

/**
 * Distinct {@link TNode} types.
 */
export type TNodeType = 'block' | 'phrasing' | 'text' | 'empty' | 'document';

export interface TNodeMethods {
  bindChildren(
    this: TNodeImpl,
    children: ReadonlyArray<TNodeImpl>,
    shouldUpdateNodeIndexes?: boolean
  ): void;
  cloneInitParams<T extends TNodeInit = TNodeInit>(
    this: TNodeImpl,
    partial?: Partial<T>
  ): T;
  isCollapsibleLeft(this: TNodeImpl): boolean;
  isCollapsibleRight(this: TNodeImpl): boolean;
  isEmpty(this: TNodeImpl): boolean;
  trimLeft(this: TNodeImpl): void;
  trimRight(this: TNodeImpl): void;
  matchContentModel(contentModel: HTMLContentModel): boolean;
  collapse(this: TNodeImpl): void;
  hasClass(className: string): boolean;
  /**
   * Collpase this node children.
   *
   * @param params
   * @returns A list of indexes to splice
   */
  collapseChildren(this: TNodeImpl): void;
  spliceChildren(this: TNodeImpl, indexes: number[]): void;
  snapshot(options?: Partial<TNodePrintOptions>): string;
  toString(): string;
  initialize(init: TNodeInit): void;
  setMarkers: SetMarkersForTNode;
}

export interface TNodeImpl<T = TNodeInit>
  extends TNodeMethods,
    Omit<TNodeShape<TNodeType>, 'children' | 'parent'> {
  __nodeIndex: number | null;
  __trimmedLeft: boolean;
  __trimmedRight: boolean;
  readonly children: ReadonlyArray<TNodeImpl>;
  readonly init: T;
  readonly parent: TNodeImpl | null;
  readonly hasWhiteSpaceCollapsingEnabled: boolean;
  readonly parentStyles: TStylesShape | null;
  readonly elementModel: HTMLElementModel<string, HTMLContentModel> | null;
  readonly contentModel: HTMLContentModel | null;
}
