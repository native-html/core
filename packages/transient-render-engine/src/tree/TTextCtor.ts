import { Text } from '../dom/dom-utils';
import HTMLContentModel from '../model/HTMLContentModel';

import compose from 'ramda/src/compose';
import { TNodeInit, TNodeImpl } from './tree-types';
import {
  normalizeWhitespaces,
  normalizeZeroWidthWhitespaces,
  removeCollapsibleAroundSegmentBreak,
  removeConsecutiveSegmentBreaks,
  removeLineBreaksAroundEastAsianDiscardSet,
  replaceSegmentBreaks
} from '../flow/text-transforms';
import TNodeCtor, { GenericTNodeCtor, Mutable } from './TNodeCtor';

export interface TTextInit extends TNodeInit {
  readonly textNode: Text;
}

export interface TTextImpl extends TNodeImpl<TTextInit> {
  readonly data: string;
  /**
   * True when tagName is defined and should be substituted with text when
   * rendering. E.g. <br> tags.
   */
  readonly isVoid: boolean;
}

const collapseWhiteSpaces = compose(
  normalizeWhitespaces,
  replaceSegmentBreaks,
  normalizeZeroWidthWhitespaces,
  removeConsecutiveSegmentBreaks,
  removeCollapsibleAroundSegmentBreak
);

const collapseWhiteSpacesWithEastAsianCharset = compose(
  normalizeWhitespaces,
  replaceSegmentBreaks,
  removeLineBreaksAroundEastAsianDiscardSet,
  normalizeZeroWidthWhitespaces,
  removeConsecutiveSegmentBreaks,
  removeCollapsibleAroundSegmentBreak
);

const TTextCtor = function TText(this: Mutable<TTextImpl>, init: TTextInit) {
  this.initialize(init);
  this.isVoid = this.elementModel ? this.elementModel.isVoid : false;
} as Function as GenericTNodeCtor<TTextInit, TTextImpl>;

//@ts-ignore
TTextCtor.prototype = new TNodeCtor('text', 'TText', {
  data: {
    get(this: TTextImpl) {
      return this.init.textNode.data;
    },
    set(this: TTextImpl, data: string) {
      this.init.textNode.data = data;
    }
  }
});

TTextCtor.prototype.matchContentModel = function matchContentModel(
  contentModel: HTMLContentModel
) {
  return (
    contentModel === HTMLContentModel.textual ||
    contentModel === HTMLContentModel.mixed
  );
};

TTextCtor.prototype.isCollapsibleLeft = function isCollapsibleLeft(
  this: TTextImpl
) {
  return (
    this.hasWhiteSpaceCollapsingEnabled &&
    !this.isEmpty() &&
    this.data[0] === ' '
  );
};

TTextCtor.prototype.isCollapsibleRight = function isCollapsibleRight(
  this: TTextImpl
) {
  return (
    this.hasWhiteSpaceCollapsingEnabled &&
    !this.isEmpty() &&
    (this.isVoid || this.data[this.data.length - 1] === ' ')
  );
};

TTextCtor.prototype.isEmpty = function isEmpty(this: TTextImpl) {
  // Only anonymous text nodes can be considered "empty"
  return this.tagName === null && !this.data.length;
};

TTextCtor.prototype.trimLeft = function trimLeft(this: Mutable<TTextImpl>) {
  if (this.isCollapsibleLeft()) {
    this.data = this.data.slice(1);
  }
};

TTextCtor.prototype.trimRight = function trimRight(this: Mutable<TTextImpl>) {
  if (this.isCollapsibleRight()) {
    this.data = this.data.substr(0, this.data.length - 1);
  }
};

TTextCtor.prototype.collapseChildren = function collapseChildren(
  this: Mutable<TTextImpl>
) {
  if (this.hasWhiteSpaceCollapsingEnabled) {
    if (this.init.context.removeLineBreaksAroundEastAsianDiscardSet) {
      this.data = collapseWhiteSpacesWithEastAsianCharset(this.data);
    } else {
      this.data = collapseWhiteSpaces(this.data);
    }
  }
  return null;
};

export default TTextCtor;

export { TTextCtor };
