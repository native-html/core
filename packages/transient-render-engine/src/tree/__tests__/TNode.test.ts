import { Element, Text } from '../../dom/dom-utils';
import TNodeCtor, { Mutable } from '../TNodeCtor';
import TTextCtor from '../TTextCtor';
import { defaultInit } from './shared';
import { TNodeImpl, TNodeInit } from '../tree-types';
import HTMLContentModel from '../../model/HTMLContentModel';

const TTest = function (this: Mutable<TNodeImpl>, init: TNodeInit) {
  this.initialize(init);
};

//@ts-ignore
TTest.prototype = new TNodeCtor('block', 'TTest');

function newTNode(init?: Partial<TNodeInit>) {
  //@ts-ignore
  return new TTest({ ...defaultInit, ...init }) as TNodeImpl;
}

describe('TNode class', () => {
  describe('isCollapsibleLeft method', () => {
    it('should return false when node has no children', () => {
      const node = newTNode();
      expect(node.isCollapsibleLeft()).toBe(false);
    });
  });
  describe('isCollapsibleRight method', () => {
    it('should return false when node has no children', () => {
      const node = newTNode();
      expect(node.isCollapsibleRight()).toBe(false);
    });
    it('should return true when node has one collapsible-right children', () => {
      const node = newTNode();
      const collapsibleChild = new TTextCtor({
        textNode: new Text('  '),
        ...defaultInit
      });
      node.bindChildren([collapsibleChild]);
      expect(node.isCollapsibleRight()).toBe(true);
    });
  });
  describe('cloneInitParams', () => {
    it('sould be immutable', () => {
      const node = newTNode();
      const nodeChildren = [newTNode()];
      node.bindChildren(nodeChildren);
      const nextNode = new TTextCtor({
        ...node.cloneInitParams(),
        parentStyles: undefined,
        styles: undefined,
        textNode: new Text('  '),
        domNode: new Element('span', { style: 'color: blue;' }),
        elementModel: null,
        nodeIndex: 10
      });
      nextNode.bindChildren([]);
      expect(node.children).toBe(nodeChildren);
      expect(node.children).toHaveLength(1);
      expect(nextNode.children).toHaveLength(0);
      expect(nextNode.type).toBe('text');
      expect(nextNode.tagName).toBe('span');
      expect(nextNode.styles.nativeTextFlow.color).toBe('blue');
    });
  });
  describe('bindChildren', () => {
    it('should not mutate prototype', () => {
      const node = newTNode();
      const nodeChildren = [newTNode()];
      node.bindChildren(nodeChildren);
      expect(TTest.prototype.children).toHaveLength(0);
    });
  });
  describe('matchContentModel', () => {
    it('should return false by default', () => {
      const node = newTNode();
      expect(node.matchContentModel(HTMLContentModel.block)).toBe(false);
    });
  });
  describe('contentModel', () => {
    it('should return null when no elementModel is available', () => {
      const node = newTNode();
      expect(node.contentModel).toBe(null);
    });
  });
  describe('snapshot', () => {
    it('should provide a JSX representation', () => {
      const node = newTNode({
        styles: {
          nativeBlockRet: { backgroundColor: 'red' },
          nativeBlockFlow: {},
          nativeTextFlow: {},
          nativeTextRet: {},
          webTextFlow: {},
          webBlockRet: {}
        }
      });
      expect(
        node.snapshot({ withNodeIndex: true, withStyles: true })
      ).toMatchSnapshot();
      expect(node.toString()).toBe(node.snapshot(/*false*/));
    });
  });
  describe('hasClass', () => {
    it('should return true when underlying DOM node has class', () => {
      const node = newTNode({
        domNode: new Element('div', {
          class: 'a b c d efg'
        })
      });
      expect(node.hasClass('efg')).toBe(true);
    });
  });
});
