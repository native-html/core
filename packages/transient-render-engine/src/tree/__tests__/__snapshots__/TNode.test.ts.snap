// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TNode class snapshot should provide a JSX representation 1`] = `"<TTest anonymous nodeIndex={0} nativeStyles={{ backgroundColor: \\"red\\" }} />"`;
